airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/InitialSyncCtidIterator.java

```diff
@@ -189,32 +208,65 @@ static List<Pair<Ctid, Ctid>> ctidQueryPlan(final Ctid startCtid,
-    LOGGER.info("Will read {} pages to get {}GB", eachStep, chunkSize);
-    final long theoreticalLastPage = relationSize / blockSize;
-    LOGGER.debug("Theoretical last page {}", theoreticalLastPage);
-    upperBound = lowerBound + eachStep;
-
-    if (upperBound > theoreticalLastPage) {
-      chunks.add(Pair.of(startCtid, null));
-    } else {
-      chunks.add(Pair.of(Ctid.of(lowerBound, startCtid.tuple), Ctid.of(upperBound, 0)));
-      while (upperBound < theoreticalLastPage) {
-        lowerBound = upperBound;
-        upperBound += eachStep;
-        chunks.add(Pair.of(Ctid.of(lowerBound, 0), upperBound > theoreticalLastPage ? null : Ctid.of(upperBound, 0)));
+    if (blockSize > 0 && chunkSize > 0 && dataSize > 0) {
+      long lowerBound = startCtid.page;
+      long upperBound;
+      final double pages = dataSize / blockSize;
+      final long eachStep = (long) pages * chunkSize;
+      LOGGER.info("Will read {} pages to get {}GB", eachStep, chunkSize);
+      final long theoreticalLastPage = relationSize / blockSize;
+      LOGGER.debug("Theoretical last page {}", theoreticalLastPage);
+      upperBound = lowerBound + eachStep;
+
+      if (upperBound > theoreticalLastPage) {
+        chunks.add(Pair.of(startCtid, null));
+      } else {
+        chunks.add(Pair.of(Ctid.of(lowerBound, startCtid.tuple), Ctid.of(upperBound, 0)));
+        while (upperBound < theoreticalLastPage) {
+          lowerBound = upperBound;
+          upperBound += eachStep;
+          chunks.add(Pair.of(Ctid.of(lowerBound, 0), upperBound > theoreticalLastPage ? null : Ctid.of(upperBound, 0)));
+        }
       }
     }
     // The last pair is (x,y) -> null to indicate an unbounded "WHERE ctid > (x,y)" query.
     // The actual last page is approximated. The last subquery will go until the end of table.
     return chunks;
   }
 
-  public PreparedStatement createCtidQueryStatement(
-                                                    final Connection connection,
+  static List<Pair<Ctid, Ctid>> ctidLegacyQueryPlan(final Ctid startCtid,
+                                                    final long relationSize,
+                                                    final long blockSize,
+                                                    final int chunkSize,
+                                                    final double dataSize,
+                                                    final int tuplesInPage) {
+
+    final List<Pair<Ctid, Ctid>> chunks = new ArrayList<>();
+    if (blockSize > 0 && chunkSize > 0 && dataSize > 0 && tuplesInPage > 0) {
+      // Start reading from one tuple after the last one that was read
+      final Ctid firstCtid = Ctid.inc(startCtid, tuplesInPage);
+      long lowerBound = firstCtid.page;
+      long upperBound;
+      final double pages = dataSize / blockSize;
+      // cap each chunk at no more than 5m tuples
+      final long eachStep = Math.min((long) pages * chunkSize,
+          5_000_000 / tuplesInPage);
```

Comment:
can you create a variable explaining the constant 5,000,000 and why it was chosen?

Reply:
It's an arbitrary large enough value that won't cause a buffer overflow on postgres due to a too large array.
Adding var